Support ElementD to be void for tma

kongroo opened the pull request:
We can use Sm90AuxStore to customize writing the D tensor (e.g. scatter the D tensor to multiple pointers). In such cases, it is necessary to remove the original writing logic of the D tensor. This PR allowes passing a nullptr for ptr_D to make storing-to-D a noop.

hwu36 commented:
@thakkarV @yzhaiustc

thakkarV commented:
Thanks a lot for your contribution, I will review your MR

thakkarV commented on include/cutlass/epilogue/collective/sm90_epilogue_tma_warpspecialized.hpp:
```diff
@@ -73,7 +74,7 @@ template <
   class CopyOpR2S_
 >
 class CollectiveEpilogue<
-    Sm90TmaWarpSpecialized<StagesC_,StagesD_,FragmentSize_,ReuseSmemC_>,
```
The way we have done this is by setting the type of `ElementD = void`. You can see this for the `ElementC` residual load as well. Can we follow the same pattern here for consistency?

kongroo replied:
I find it not easy to make both C and D void because ElementD is used in many places such as construction tensors and calculating tile shapes in the builder.

thakkarV commented on include/cutlass/epilogue/collective/sm90_epilogue_tma_warpspecialized.hpp:
```diff
@@ -600,7 +612,14 @@ class CollectiveEpilogue<
-        copy(tiled_r2s, tRS_rD, tRS_sD(_,_,_,store_pipe_producer_state.index()));
+        [&]() -> void {
+          if constexpr (DispatchPolicy::EnableNullptrD) {
+            if (params.is_nullptr_d) {
```
is a dynamic branch needed here? looks like you always check against the template parameter only. So we can totally avoid any dynamic checks within the kernel which will ensure no perf regressions.

kongroo replied:
I followed the code pattern in Sm90AuxStore. I think there are two possible implementations here. 1. use EnableNullptrD = true as default and dynamic check; 2. use DisableWriteD = false as default and constant check; I think It makes sense to prefer the second one for D.

kongroo replied:
I have updated to use static branches. Could you please take a look?

thakkarV commented on include/cutlass/epilogue/collective/sm90_epilogue_tma_warpspecialized.hpp:
```diff
@@ -209,6 +210,7 @@ class CollectiveEpilogue<
     typename FusionCallbacks::Params thread{};
     TMA_C tma_load_c;
     TMA_D tma_store_d;
+    bool is_nullptr_d = false;
```
since you are only using the template parameter, this is not needed.

thakkarV commented on include/cutlass/epilogue/dispatch_policy.hpp:
```diff
@@ -131,13 +131,15 @@ template<
   int FragmentSize_,
-  bool ReuseSmemC_
+  bool ReuseSmemC_,
+  bool EnableNullptrD_ = true
```
See comment above about using the element D to detect this instead of adding a new parameter here.

yzhaiustc commented on include/cutlass/epilogue/dispatch_policy.hpp:
```diff
@@ -131,13 +131,15 @@ template<
   int FragmentSize_,
-  bool ReuseSmemC_
+  bool ReuseSmemC_,
+  bool EnableNullptrD_ = true
```
Nit: should the `EnableNullptrD_` be `false` by default?

kongroo replied:
Updated

yzhaiustc replied:
Thank you. Cleared on my side.

richardmcai commented:
if you know that storing D will be disabled then adding support for `void` D is probably the better way here, instead of adding these runtime checks. You can follow how `is_source_supported` is used for reference, you could create a similar `is_destination_supported` variable for void D

thakkarV commented:
@richardmcai thoughts on this MR now?

richardmcai commented:
> @richardmcai thoughts on this MR now?

It should be fine, a couple points
* It's not clear to me how this feature is going to be exposed to the user since there are no builder modifications. Is this collective being instantiated manually?
* Ultimately this flag should be deprecated in favor of supporting void D to maintain consistency with void C behavior. Fine with accepting it for now with the understanding that it may be deprecated in a future release (with appropriate prior notice or compile warnings)

kongroo commented:
> > @richardmcai thoughts on this MR now?
> 
> It should be fine, a couple points
> 
> * It's not clear to me how this feature is going to be exposed to the user since there are no builder modifications. Is this collective being instantiated manually?
> * Ultimately this flag should be deprecated in favor of supporting void D to maintain consistency with void C behavior. Fine with accepting it for now with the understanding that it may be deprecated in a future release (with appropriate prior notice or compile warnings)

Yes we manually construct the collective. Should I add some comments to suggest that NoWriteD maybe remoted if D=void is supported?

hwu36 commented:
@kongroo , would you please rebase your code?  sm90 epilogue code changed quite a bit.  there are conflicts.  i tried to solve it myself, but ends up with mismatch results.

kongroo commented:
@hwu36 I just rebased the code. (but haven't checked the correctness yet)

kongroo commented:
@hwu36 It looks like version 3.4 has been delayed? Do you know when it will get ready? I'll rebase then

kongroo commented:
@hwu36 @richardmcai @yzhaiustc @thakkarV  I have added support for passing void to ElementD, and I've also added a unit test, which compares the output of writing D using normal kernel with using AuxStore and void ElementD. Could you please review again?
